Lib/ufo2ft/featureWriters/gdefFeatureWriter.py

```diff
@@ -14,34 +12,41 @@ class GdefFeatureWriter(BaseFeatureWriter):
     """
 
     tableTag = "GDEF"
+    features = frozenset(["GlyphClassDefs", "LigatureCarets"])
+    insertFeatureMarker = None
 
     def setContext(self, font, feaFile, compiler=None):
-        ctx = self.context = SimpleNamespace(
-            font=font,
-            feaFile=feaFile,
-            compiler=compiler,
-            todo=set(),
-        )
-        # skip if a GDEF is in the features
-        if ast.findTable(self.context.feaFile, "GDEF") is not None:
-            return ctx
+        ctx = super().setContext(font, feaFile, compiler=compiler)
+
+        ctx.gdefTableBlock = ast.findTable(self.context.feaFile, "GDEF")
+        if ctx.gdefTableBlock:
+            for fea in ctx.gdefTableBlock.statements:
+                if isinstance(fea, ast.GlyphClassDefStatement):
+                    ctx.todo.remove("GlyphClassDefs")
+                elif isinstance(fea, ast.LigatureCaretByIndexStatement) or isinstance(
+                    fea, ast.LigatureCaretByPosStatement
+                ):
+                    ctx.todo.remove("LigatureCarets")
 
         ctx.orderedGlyphSet = self.getOrderedGlyphSet()
-        ctx.ligatureCarets = self._getLigatureCarets()
-        ctx.openTypeCategories = self.getOpenTypeCategories()
-        ctx.todo.update({"GlyphClassDefs", "LigatureCarets"})
+
+        if "GlyphClassDefs" in ctx.todo:
+            ctx.openTypeCategories = self.getOpenTypeCategories()
+        if "LigatureCarets" in ctx.todo:
+            ctx.ligatureCarets = self._getLigatureCarets()
 
         return ctx
 
     def shouldContinue(self):
-        if not self.context.todo:
+        todo = self.context.todo
+        if not todo:
             return super().shouldContinue()
 
-        if not self.context.ligatureCarets:
-            self.context.todo.remove("LigatureCarets")
+        if "LigatureCarets" in todo and not self.context.ligatureCarets:
+            todo.remove("LigatureCarets")
```

Comment:
I think you can move this todo.remove() to the setContext method, immediately after you try to get them but found none. This way you can leave the base class' shouldContinue untouched

Reply:
OK. `setContext()` does this instead now.